src/main/java/com/limechain/runtime/hostapi/OffchainHostFunctions.java

```diff
@@ -214,4 +238,249 @@ public int extOffchainRandomSeed() {
+            if (response[0] == HttpResponseType.FINISHED) {
+                byte[] data = requests.getResponseBody(requestId);
+                hostApi.writeDataToMemory(data, bufferPointer);
+
+                byte[] resultValue = scaleEncodeIntResult(data.length);
+                return hostApi.writeDataToMemory(resultValue);
+            } else {
+                switch (response[0]) {
+                    case INVALID_ID:
+                        return hostApi.writeDataToMemory(HttpResponseType.INVALID_ID.scaleEncodedResult());
+                    case DEADLINE_REACHED:
+                        return hostApi.writeDataToMemory(HttpResponseType.DEADLINE_REACHED.scaleEncodedResult());
+                    case IO_ERROR:
+                        return hostApi.writeDataToMemory(HttpResponseType.IO_ERROR.scaleEncodedResult());
+                    default:
+                        return null;
+                }
+            }
+        } catch (InvalidRequestId e) {
+            return hostApi.writeDataToMemory(HttpResponseType.INVALID_ID.scaleEncodedResult());
+        } catch (SocketTimeoutException e) {
+            log.log(Level.WARNING, e.getMessage(), e.getStackTrace());
+            return hostApi.writeDataToMemory(HttpResponseType.DEADLINE_REACHED.scaleEncodedResult());
+        } catch (IOException e) {
+            log.log(Level.WARNING, e.getMessage(), e.getStackTrace());
+            return hostApi.writeDataToMemory(HttpResponseType.IO_ERROR.scaleEncodedResult());
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new ThreadInterruptedException(e);
+        }
+    }
+
+    byte[] scaleEncodeHeaders(Map<String, List<String>> headers) throws IOException {
+        List<Pair<String, String>> pairs = new ArrayList<>(headers.size());
+        for (Map.Entry<String, List<String>> entry : headers.entrySet()) {
+            // NOTE: Not sure if a comma is the right separator
+            String value = String.join(",", entry.getValue());
+            pairs.add(new Pair<>(entry.getKey(), value));
+        }
+        return ScaleUtils.Encode.encode(pairs, String::getBytes, String::getBytes);
+    }
+
+    int timeoutFromDeadlineAndTimestamp(RuntimePointerSize deadlinePointer, long timestamp) {
+        byte[] deadlineBytes = hostApi.getDataFromMemory(deadlinePointer);
+        return new ScaleCodecReader(deadlineBytes)
+                .readOptional(new UInt64Reader())
+                .map(deadline -> deadline.longValue() - timestamp)
+                .map(Long::intValue)
+                .orElse(0);
+    }
+
+    int[] decodeRequestIdArray(byte[] encodedRequestIds) {
+        ScaleCodecReader reader = new ScaleCodecReader(encodedRequestIds);
+        int length = reader.readCompactInt();
+        int[] requestIds = new int[length];
+        for (int i = 0; i < length; i++) {
+            requestIds[i] = (int) reader.readUint32();
```

Comment:
Is it safe to cast u32 to int here? The upper half of u32 values overflow int

Reply:
Overflowing would make the integer negative but since we store the ids in a Map it would still find the corresponding request

Reply:
This [https://github.com/paritytech/polkadot-sdk/blob/9f5221cc2f7f10adfd22b293ceed060617468936/substrate/primitives/core/src/offchain/mod.rs#L101C1-L101C1](struct) is used for containing the Ids of the requests and is of type i16 so we won't get to a point where we overflow.